rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,755 @@
+until the resource exists in the local backend or for replication to complete to the leaf backend.
+
+```proto
+package teleport.mfa.v1;
+
+// MFAService defines the Multi-Factor Authentication (MFA) service. While this service is currently focused on
+// user sessions, new MFA related RPCs should be added here instead of the AuthService, to maintain a clear separation
+// of concerns instead of further bloating the AuthService.
+service MFAService {
+  // CreateChallenge creates an MFA challenge that is tied to a user session.
+  rpc CreateChallenge(CreateChallengeRequest) returns (CreateChallengeResponse);
+  // ValidateChallenge validates the MFA challenge response for a user session and stores the validated response in the
+  // backend.
+  rpc ValidateChallenge(ValidateChallengeRequest) returns (ValidateChallengeResponse);
+  // ReplicateValidatedMFAChallenge replicates a validated MFA challenge to a leaf cluster for retrieval during SSH session
+  // establishment. It is a NOOP when used in the root cluster. It is called by the reverse tunnel server when a
+  // validated challenge is created for a leaf cluster.
+  rpc ReplicateValidatedMFAChallenge(ReplicateValidatedMFAChallengeRequest) returns (ValidatedMFAChallenge);
+  // VerifyValidatedMFAChallenge verifies a previously validated MFA challenge response for a user session.
+  // If the challenge does not yet exist, this method will block until the resource appears or until the timeout is
+  // reached. The payload is used to verify the challenge is tied to the correct user session. If the verification is
+  // successful, the MFA device used for authentication is returned. If the verification fails, an error is returned.
+  rpc VerifyValidatedMFAChallenge(VerifyValidatedMFAChallengeRequest) returns (VerifyValidatedMFAChallengeResponse);
+}
+
+// SessionIdentifyingPayload contains a value that uniquely identifies a user's session.
+// It must be computed by the client from session state and is used to bind MFA challenges to specific sessions.
+message SessionIdentifyingPayload {
+  oneof payload {
+    teleport.ssh.v1.SessionPayload ssh_session = 1;
+  }
+}
+
+// CreateChallengeRequest is the request message for CreateChallenge.
+message CreateChallengeRequest {
+  // payload is a value that uniquely identifies the user's session. When VerifyValidatedMFAChallenge is called, the server
+  // will verify it matches the payload supplied to CreateChallengeRequest.
+  SessionIdentifyingPayload payload = 1;
+  // sso_client_redirect_url should be supplied if the client supports SSO MFA checks. If unset, the server will only
+  // return non-SSO challenges.
+  string sso_client_redirect_url = 2;
+  // proxy_address is the proxy address that the user is using to connect to the Proxy. When using SSO MFA, this address
+  // is required to determine which URL to redirect the user to when there are multiple options.
+  string proxy_address = 3;
```

Comment:
I understand that this matches the existing API but "proxy_address" is _really_ not specific enough.

```suggestion
  string proxy_address_for_sso = 3;
```